### PR TITLE
Add keyring snap removal confirmation

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This feature adds an extra confirmation prior to the removal of a keyring snap. The confirmation will display the accounts that are associated to the snap and warn the user that these accounts will be deleted. The confirmation also includes user friction to type in the snap name prior to the removal. 

Related to: https://github.com/MetaMask/accounts-planning/issues/13

## **Manual testing steps**

1. Start metamask flask
2. Install [Snap Simple Keyring version 0.2.2](https://metamask.github.io/snap-simple-keyring/0.2.2/)
3. Create an account 
4. On the metamask extension, click on settings
5. Click on snaps
6. Select the Snap Simple Keyring
7.  Click on `Remove Metamask Snap Simple Keyring`
8.  Click `Continue`
9.  Enter the snap name and click `Remove Snap`

### **Before**

remove snap confirmation 5

### **After**

remove snap confirmation 1
remove snap confirmation 2

Success Case:
remove snap confirmation 3

Failure Case:
remove snap confirmation 4

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by brad-decker — DISMISSED** [review]
🎉

**Comment by coreyjanssen** [comment]
Can we change the button `Nevermind` to `Cancel` please?  And please capitalize `Snap` in the screen title and body of the warning messages. @montelaidev

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by montelaidev** [comment]
@SocketSecurity ignore webextension-polyfill-ts@0.25.0

This is imported by [@metamask/accounts-controller@1.0.0](https://socket.dev/npm/package/@metamask/accounts-controller/overview/1.0.0), [@metamask/eth-snap-keyring@0.2.3](https://socket.dev/npm/package/@metamask/eth-snap-keyring/overview/0.2.3), [@metamask/keyring-api@0.2.7](https://socket.dev/npm/package/@metamask/keyring-api/overview/0.2.7), [@metamask/providers@10.2.1](https://socket.dev/npm/package/@metamask/providers/overview/10.2.1), [@metamask/snaps-controllers@2.0.0](https://socket.dev/npm/package/@metamask/snaps-controllers/overview/2.0.0)

**Comment by AlexJupiter** [comment]
From what I can see, I think this ticket is good enough to go. I understand there's some improvements to the copy needed, but I'd like to advocate we do that in a follow on ticket to not slow this main one down.

File: ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                      key="keyringSnapRemoveConfirmation2"
+                      fontWeight={FontWeight.Bold}
+                      as="span"
+                    >
+                      {snap.manifest.proposedName}
+                    </Text>,
+                  ])}
+                </Text>
+                {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
+                {/* @ts-ignore TODO: fix TextField props */}
```
Comment (hmalik88) [inline ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:138]: what's missing here?

File: ui/pages/settings/snaps/view-snap/view-snap.js [hunk 2]
```diff
@@ -18,26 +18,42 @@ import SnapAuthorshipExpanded from '../../../../components/app/snaps/snap-author
...
 import { DelineatorType } from '../../../../helpers/constants/snaps';
 
+///: BEGIN:ONLY_INCLUDE_IN(keyring-snaps)
+const KeyringSnapRemovalResultStatus = {
+  Success: 'success',
+  Failed: 'failed',
+  None: 'none',
+};
+///: END:ONLY_INCLUDE_IN
+
```
Comment (hmalik88) [inline ui/pages/settings/snaps/view-snap/view-snap.js:56]: Can we move this enum to a different file?

File: ui/pages/settings/snaps/view-snap/view-snap.js [hunk 3]
```diff
@@ -190,13 +214,48 @@ function ViewSnap() {
             </Text>
           </Button>
           <SnapRemoveWarning
-            isOpen={isShowingRemoveWarning}
+            isOpen={isShowingRemoveWarning && !isKeyringSnap}
```
Comment (hmalik88) [inline ui/pages/settings/snaps/view-snap/view-snap.js:217]: is the second condition needed? isKeyringSnap is not set to `true` otherwise.

Comment (montelaidev) [reply ui/pages/settings/snaps/view-snap/view-snap.js:217]: The second condition is to show the original snap removal confirmation if its not a keyring snap

File: ui/pages/settings/snaps/view-snap/view-snap.js (hunk 2 above)
Comment (montelaidev) [reply ui/pages/settings/snaps/view-snap/view-snap.js:51]: Sure, i've made the change in 899aba2246a6d420857c7a4de58584e5569a7604

File: ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx (hunk 1 above)
Comment (montelaidev) [reply ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:138]: Currently all the TextField props are all required, and we would have to change it to this 

`<TextField
                  marginBottom={4}
                  value={confirmationInput}
                  onChange={(e: { target: { value: string } }) => {
                    setConfirmationInput(e.target.value);
                    setConfirmedRemoval(
                      validateConfirmationInput(e.target.value),
                    );
                  }}
                  error={error}
                  inputProps={{
                    'data-testid': 'remove-snap-confirmation-input',
                  }}
                  type="text"
                  autoComplete={undefined}
                  autoFocus={undefined}
                  className={undefined}
                  defaultValue={undefined}
                  disabled={undefined}
                  id={undefined}
                  inputRef={undefined}
                  startAccessory={undefined}
                  endAccessory={undefined}
                  maxLength={undefined}
                  name={undefined}
                  onBlur={undefined}
                  onClick={undefined}
                  onFocus={undefined}
                  placeholder={undefined}
                  readOnly={undefined}
                  required={undefined}
                  testId={undefined}
                />`

**Review by gantunesr — DISMISSED** [review]
LGTM

File: ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx (hunk 1 above)
Comment (hmalik88) [reply ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:138]: Gotcha

**Review by hmalik88 — DISMISSED** [review]

**Comment by plasmacorral** [comment]
Tested on Mac Sonoma 14.0 with Chrome 117 primarily with a spot check on Firefox 118.  Blocking concern captured in [76](https://github.com/MetaMask/accounts-planning/issues/76). With further opportunities for improvement captured in [73](https://github.com/MetaMask/accounts-planning/issues/73), [77](https://github.com/MetaMask/accounts-planning/issues/77) and [78](https://github.com/MetaMask/accounts-planning/issues/78).

If we are making any further changes here, I will request that we also merge the latest develop so that this feature can be tested with the changes introduced in PR [20684](https://github.com/MetaMask/metamask-extension/pull/20684).

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by plasmacorral** [comment]
> @plasmacorral following up on this from our conversation yesterday, all the tickets your reference above seem either low priority or unrelated to this ticket. Is it possible to move this forwards? cc. @gantunesr

I was waiting for more information around issue [76](https://github.com/MetaMask/accounts-planning/issues/76), but looks like we now have an additional PR in draft. E2e is not passing here, and we need dev approvals once again. @AlexJupiter

**Review by gantunesr — DISMISSED** [review]

**Review by hmalik88 — DISMISSED** [review]

**Comment by plasmacorral** [comment]
I am seeing an issue with commit 0e533d88c4136bc77559da9ca74fd1e13d520182 that I am unable to remove the Snap if there are no snap accounts present.  The `Remove MetaMask Simple Snap Keyring` CTA does nothing.  This can be observed running a snap locally, or when SSK is installed from snaps grid.

Recording: https://recordit.co/EDY9rIWNbw

Mac Sonoma 14
Chrome  117.0.5938.92
Observed on a dist build w/ `yarn dist --build-type flask`

I will also note that the back arrow now functions from the Remove Snap- confirm page.  However, on the initial Remove Snap where the public address is displayed, the back arrow does nothing.  I expect that it should return the user to settings>snaps>snap detail

**Review by Ameralameri — APPROVED** [review]
Approved

**Review by hmalik88 — APPROVED** [review]

**Comment by danroc** [comment]
If the snap takes more time to process the account deletion, this happens:

https://github.com/MetaMask/metamask-extension/assets/68558152/9c679861-7567-478b-a7f8-7d25b6223130

To make it more apparent, I added a sleep in the `deleteAccount` method inside the snap:

```ts
await new Promise((r) => setTimeout(r, 2000));
```

**Review by gantunesr — APPROVED** [review]
LGTM. Lets just make sure to address @danroc's comment in this or a follow up PR

**Comment by montelaidev** [comment]
Video of the new behaviour
https://github.com/MetaMask/metamask-extension/assets/96463427/08953cd0-a7b0-427d-b511-6b97f4adc6fe

**Pull request merged**